api/apiclient.go

```diff
@@ -605,13 +601,16 @@ func dialAPI(info *Info, opts0 DialOpts) (*dialResult, error) {
 
 // gorillaDialWebsocket makes a websocket connection using the
 // gorilla websocket package.
-func gorillaDialWebsocket(ctx context.Context, urlStr string, tlsConfig *tls.Config) (jsoncodec.JSONConn, error) {
+func gorillaDialWebsocket(ctx context.Context, urlStr string, tlsConfig *tls.Config, ipAddr string) (jsoncodec.JSONConn, error) {
 	// TODO(rogpeppe) We'd like to set Deadline here
 	// but that would break lots of tests that rely on
 	// setting a zero timeout.
 	netDialer := net.Dialer{}
 	dialer := &websocket.Dialer{
 		NetDial: func(netw, addr string) (net.Conn, error) {
+			if ipAddr != "" {
```

Comment:
This sort of thing always feels bad to me.
We've done "foo.NetDial(X)" but under the covers we actually force that to be "net.Dial(Y)" and we never even check that 'addr' matches the object we're trying to substitute.
If we wanted to dial Y, why didn't we do so from the beginning?
I'm guessing this is again the hostname vs IP thing, but I don't quite understand why we need to replace DNS.